### ISSUE TITLE
Why ranx is too slow in this simple example?

### ISSUE THREAD

**Issue by celsofranssa** [AmenRa/ranx#23]
```python
from ranx import Qrels
from ranx import Run
from ranx import evaluate

qrels_dict = {
    "text_1": {
        "label_1": 1
    },
    
    "text_2":{
        "label_2": 1,
    }
}

qrels = Qrels(qrels_dict, name="testing")



run_dict = {
    "text_1": {
        "label_1": 1,
        "label_2": 0.9,
        "label_3": 0.8,
        "label_4": 0.7,
        "label_5": 0.6,
        "label_6": 0.5,
        "label_7": 0.4,
        "label_8": 0.3,
        "label_9": 0.2,
        "label_10": 0.1,
    },
    
    "text_2": {
        "label_1": 0.9,
        "label_2": 1,
        "label_3": 0.8,
        "label_4": 0.7,
        "label_5": 0.6,
        "label_6": 0.5,
        "label_7": 0.4,
        "label_8": 0.3,
        "label_9": 0.2,
        "label_10": 0.1,
    },
}

run = Run(run_dict, name="bm25")

evaluate(qrels, run, ["mrr@1", "mrr@5", "mrr@10"])
```
```python
CPU times: user 55.2 s, sys: 467 ms, total: 55.7 s
Wall time: 56.8 s
```

This behavior can be reproduced in this [Google Colab Notebook 1](https://colab.research.google.com/drive/1nakaFZYsbiyOfRBuKO8TYI7VtYLQQQER?usp=sharing) and also in this [Google Colab Notebook 2](https://colab.research.google.com/drive/1sHuAWov1oH3Z8gAk-ujEvNTgtXtP4T3Z?usp=sharing) (spent time per steps).

**Comment by celsofranssa** [comment]
Wow, it is taking forever in a real example with about 60k queries!
Am I missing something?

**Comment by AmenRa** [comment]
Hi, and thanks for your interest in `ranx`!

I think you are missing that all [`numba`](https://numba.pydata.org)-based `ranx` functions need to be compiled the first time you use them (_there is a disclaimer on top of each of my notebooks about that_).
Also, Google Colab is very slow in compiling them.

If you run your notebook again (without reloading Colab), you should notice an extremely lower computation time.

Unfortunately, I suspect you must recompile `ranx` functionalities every time you start a new Colab instance.
On your local machine, the compiled functions should be automatically stored for future usage by numba.

If you are stuck with Colab, I suggest you compile the functions with toy examples before using them with real-world data.
You should be absolutely fine then with 60k queries, especially with `MRR`, which is very optimized.

On my local machine, a four years old MBP, I get these execution times for 1M queries with 100 results each:

```python
%%time
evaluate(qrels, run, ["mrr@1", "mrr@5", "mrr@10"])
```

```python
CPU times: user 11.6 s, sys: 38.6 ms, total: 11.6 s
Wall time: 1.05 s
```

```python
%%time
evaluate(qrels, run, ["map", "mrr", "ndcg"])
```

```python
CPU times: user 25.8 s, sys: 73.8 ms, total: 25.9 s
Wall time: 2.3 s
```

Hope this answer your question.

Please, consider giving `ranx` a start if you like it!

Best,

Elias

**Comment by celsofranssa** [comment]
Hi @AmenRa,

That was the case. Thank you for your quick answer.
And it will be a pleasure to give ranx a star.

**Comment by AmenRa** [comment]
@milyenpabo Have you already tried with a dummy Qrels?
Could you please post a sample of your specific Qrels without any modification?

**Comment by milyenpabo** [comment]
Ok, I'm reading up on Numba a bit, and I think I found the root cause.

1. I found an option to disable JIT compilation:

https://numba.readthedocs.io/en/stable/user/troubleshoot.html#disabling-jit-compilation

I tried it, and my test program gets significantly faster:

```
[INFO] 2023-11-14 20:50:48.255 generated new fontManager
[INFO] 2023-11-14 20:50:48.785 Crearting a small Qrels object.
[INFO] 2023-11-14 20:50:48.785 Qrels object created in 0.00 seconds.
```

So, this is good for a quick fix. Although, this way I guess we lose the benefits of Numba for larger eval sets? So, a less pressing question is: is there a way to compile once and reuse for subsequent runs?

2. I found the `cache=true` option in the Numba docs, precisely to allow for reusing the numba-compiled code:

https://numba.readthedocs.io/en/stable/user/faq.html#there-is-a-delay-when-jit-compiling-a-complicated-function-how-can-i-improve-it

Checking `ranx`, it seems it uses the `cache=True` option (most of the time):

https://github.com/search?q=repo%3AAmenRa%2Franx+jit&type=code

3. I realized I left out a seemingly minor detail from my previous post: I run the program in a Docker container. So, it might be that every time the program terminates and the container is stopped and removed, I just lose the Numba cache with it? After a bit of digging, I found a way to specify the cache directory:

https://numba.readthedocs.io/en/stable/reference/envvars.html#numba-envvars-caching

So, I tried setting the `NUMBA_CACHE_DIR` to point to a directory which is mapped from the host. Unfortunately it does not work. I don't see any files created by Numba, and hence I don't see any speed-up on repeated runs.

So, the question is narrowed down to: how can I get ranx/Numba to use the cache?
